### PR TITLE
Additional Heartbeat Reload Tests

### PR DESCRIPTION
**Pull request by andrewvc**
This PR attempts to improve the testing and testability of heartbeat in the following ways:

1. Adds a mocked test of the `Monitor` type
2. Minimizes the required mockable area by introducing a `PipelineConnector` type that is easier to mock than a `Pipeline`.
3. Introduces a number of useful mocks for future tests potentially.

### CODE REVIEW

**Comment by andrewvc** [comment]
This is now ready for review

File: heartbeat/monitors/monitor_test.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	mon, err := newMonitor(serverMonConf, reg, pipelineConnector, sched, false)
+	require.NoError(t, err)
+
+	mon.Start()
+	defer mon.Stop()
+
+	require.Equal(t, 1, len(pipelineConnector.clients))
+	pcClient := pipelineConnector.clients[0]
+
+	time.Sleep(time.Duration(100 * time.Millisecond))
```
Comment (ruflin) [inline heartbeat/monitors/monitor_test.go]: Sleeps in tests ;-)

**Comment by andrewvc** [comment]
@ruflin fixed and improved in 35a9170

The issue was that the mock impl was doing a copy into a zero size slice with the right capacity, instead of an empty slice of the right capacity.

I also improved the timeout code to exit early by spinning on the condition. Wouldn't mind some review on this, my understanding is that the go scheduler is not safe to spin in this situation and still make forward-progress without explicit sleeps in it, even on a single core system. Is that correct? I added microsecond sleeps in the loop to make sure we yield. 

Looking at https://github.com/golang/go/issues/10958 seems to confirm that aspect of the go scheduler behavior.

**Comment by ruflin** [comment]
It seems on Jenkins Mac build this causes a problem in libbeat: https://beats-ci.elastic.co/job/elastic+beats+pull-request+multijob-darwin/483/beat=libbeat,label=macosx/console I can't really see how this is related to the interface change in libbeat TBH. Perhaps it's flaky but it's not one I have seen before.

**Pull request merged**